### ISSUE TITLE
EoD API Shutdown Causes Crash

### ISSUE THREAD

**Issue by romen-h** [blish-hud/Blish-HUD#619]
I picked the worst possible time to try Blish-HUD for the first time.

Currently Anet has disabled the API to avoid leaking information.

A side-effect of this is that Blish-HUD appears to be crashing when it encounters the 503 response that the API replies with right now.

I can reproduce a crash by adding my API key to Blish-HUD on a freshly installed copy.

I realize this problem will fix itself when the API comes back online; But for long term stability I would suggest fixing something so it gracefully handles these unexpected API responses.

Here is a log from one of the crashes:

```
Blish HUD v0.11.7+Branch.tags-v0.11.7.Sha.0c709c76cb502d310f2964dca92ce770e533dae2
23:44:55.2153 | INFO  | Blish_HUD.Program | Running Build "Branch.tags-v0.11.7.Sha.0c709c76cb502d310f2964dca92ce770e533dae2"
23:44:55.2193 | INFO  | Blish_HUD.Program | Launched from "F:\GameTools\gw2\Blish.HUD.0.11.7" with args "".
23:44:57.1203 | INFO  | Blish_HUD.Gw2WebApiService | Could not find registered API key associated with character "Víridia"
23:44:57.1543 | WARN  | Blish_HUD.Contexts.FestivalContext | Failed to query Guild Wars 2 API.
Gw2Sharp.WebApi.Exceptions.ServiceUnavailableException: <h1>Status Code 503</h1><p>Temporarily disabled to avoid classified End of Dragons information escaping from the vault of forbidden spoilers. See you in Cantha!</p>
   at Gw2Sharp.WebApi.Middleware.ExceptionMiddleware.<>c__DisplayClass1_0.<<OnRequestAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.RequestSplitterMiddleware.<>c__DisplayClass4_0.<<OnRequestAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.CacheMiddleware.<RequestGetAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Caching.BaseCacheMethod.<>c__DisplayClass4_0.<<GetOrUpdateAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.CacheMiddleware.<OnEndpointRequestAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Http.WebApiRequest.<ExecuteAsync>d__10`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.V2.Clients.BaseEndpointBulkAllClient`2.<GetAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Blish_HUD.Contexts.FestivalContext.<GetFestivalsFromGw2Api>d__9.MoveNext() in D:\a\Blish-HUD\Blish-HUD\Blish HUD\GameServices\Contexts\FestivalContext.cs:line 153

23:44:57.1573 | INFO  | Blish_HUD.GameIntegration.ClientTypeIntegration | Detected Guild Wars 2 client to be the Standard version.
23:44:57.1573 | INFO  | Blish_HUD.Contexts.FestivalContext | Active festival(s): ""
23:44:57.1573 | INFO  | Blish_HUD.Modules.Pkgs.StaticPkgRepoProvider | Getting module list...
23:46:01.3743 | FATAL | Blish_HUD.DebugService | Blish HUD encountered a fatal crash!
Gw2Sharp.WebApi.Exceptions.ServiceUnavailableException: <h1>Status Code 503</h1><p>Temporarily disabled to avoid classified End of Dragons information escaping from the vault of forbidden spoilers. See you in Cantha!</p>
   at Gw2Sharp.WebApi.Middleware.ExceptionMiddleware.<>c__DisplayClass1_0.<<OnRequestAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.RequestSplitterMiddleware.<>c__DisplayClass4_0.<<OnRequestAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.CacheMiddleware.<RequestGetAsync>d__8.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Caching.BaseCacheMethod.<>c__DisplayClass4_0.<<GetOrUpdateAsync>g__ExecAsync|0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Middleware.CacheMiddleware.<OnEndpointRequestAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.Http.WebApiRequest.<ExecuteAsync>d__10`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Gw2Sharp.WebApi.V2.Clients.BaseEndpointBlobClient`1.<GetAsync>d__1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Blish_HUD.Gw2WebApi.UI.Views.RegisterApiKeyView.<UpdateTokenDetails>d__25.MoveNext() in D:\a\Blish-HUD\Blish-HUD\Blish HUD\GameServices\Gw2WebApi\UI\Views\RegisterApiKeyView.cs:line 326
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Blish_HUD.Gw2WebApi.UI.Views.RegisterApiKeyView.<CheckToken>d__24.MoveNext() in D:\a\Blish-HUD\Blish-HUD\Blish HUD\GameServices\Gw2WebApi\UI\Views\RegisterApiKeyView.cs:line 302
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
   at System.Threading.ThreadPoolWorkQueue.Dispatch()

```

**Comment by dlamkins** [comment]
Thanks for the trace.  This is something we're definitely aware of with the discussion in our Discord server.  The odd thing is that from what I can tell, the exception should have been caught.

@Archomeda any ideas how a `ServiceUnavailableException` could have slipped through our try/catches here?

![image](https://user-images.githubusercontent.com/1950594/155929975-3bfa5ed5-d9e1-45a7-85cd-5dc98624824f.png)

https://github.com/blish-hud/Blish-HUD/blob/0c709c76cb502d310f2964dca92ce770e533dae2/Blish%20HUD/GameServices/Gw2WebApi/UI/Views/RegisterApiKeyView.cs#L289-L321

You can see we catch some explicit exceptions to help ensure that the message we return back to the user is a bit more accurate about the issue, but [ServiceUnavailableException](https://github.archomeda.eu/Gw2Sharp/master/api/Gw2Sharp.WebApi.Exceptions.ServiceUnavailableException.html) appears to inherit from `UnexpectedStatusException`.  You can see my intention here was to use the parent exception as a catch type, but that doesn't seem to have worked here.

Earlier in the day when the API was returning 418, it did appear to be working, though.  🤔 I suspect I'm missing something simple.

**Comment by Archomeda** [comment]
@dlamkins Yeah, this is an odd one that I've been wanting to fix in a v2.0. The ServiceUnavailableException inherits from the generic variant of UnexpectedStatusException, a.k.a. `UnexpectedStatusException<ErrorObject>`.

[It's in the pipeline for a v2.0 to be simplified.](https://github.com/Archomeda/Gw2Sharp/blob/2.0/CHANGELOG.md#fixes)

**Comment by dlamkins** [comment]
Ahhh dang it.  That explains it.  Thanks.  👍

**Comment by de6orah** [comment]
I arrived at this page to see why my lovely Blish HUD didn't want to work yesterday and found the answer: thank you. I have nothing helpful to add to this discussion but did want to take this opportunity to thank the Blish HUD team for all their hard work. It really is a wonderful tool for GW2 and I really appreciate it, and the team.

**Comment by dlamkins** [comment]
Alright this is now resolved in #621 and will be available in the next release.

@de6orah and thank you for your comments.  They're very much appreciated.  :smile:
